### PR TITLE
Replace the buffer_t host_dirty/dev_dirty fields with flags

### PR DESCRIPTION
**Pull request by steven-johnson**
Adding #pragma pack turned out to have an unpleasant side effect: it
removed the alignment restriction on the entire buffer_t struct. On
some platforms this can cause misaligned loads. Let’s avoid this
entirely by replacing the fields at the end with an int32 bit mask.

(Also remove a few more rogue definitions of buffer_t.)

### CODE REVIEW

File: src/runtime/HalideRuntime.h [hunk 1]
```diff
@@ -338,22 +337,37 @@ typedef struct buffer_t {
...
+};
 
 #endif
 
+inline bool halide_get_buffer_flag(const buffer_t* buf, halide_buffer_flag_bits_t flag) {
+    return (buf->flags & flag) != 0;
+}
+
+inline void halide_set_buffer_flag(buffer_t* buf, halide_buffer_flag_bits_t flag, bool set) {
+    if (set)
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: We typically use braces on ifs split over multiple lines.

**Comment by abadams** [comment]
I get a segfault in test_memoize. That code is sensitive to sizeof(buffer_t). However it says it's 72 bytes, which is correct. Investigating...

**Comment by steven-johnson** [comment]
test_memoize runs properly on my local Mac.

**Comment by abadams** [comment]
... except I'm building it in 32-bit mode, and pragma pack shrinks it down to 68 bytes, which is the actual correct number for 32 bit. sigh.

**Comment by abadams** [comment]
(and pragma pack makes it work again)

**Comment by steven-johnson** [comment]
#pragma pack needs to die in a fire. where is the extra padding being added?

**Comment by abadams** [comment]
I'll print out the offset of each member...

**Comment by abadams** [comment]
At the end, after the flags, in order to bring the struct up to a multiple of 8 bytes. Presumably so that the dev field can be aligned in an array of buffer_ts

**Comment by steven-johnson** [comment]
Ahhh... of course, because you might allocate buffer_t[MAXINT]. :-/

Willfix.

**Comment by abadams** [comment]
Not pretty, but I guess host could be in a union with a uint64_t, so that the host pointer always takes up 8 bytes and the whole struct always takes 72.

**Comment by steven-johnson** [comment]
That would require touching a whoooooole lotta code. Let's not go there just yet.

**Comment by steven-johnson** [comment]
We could always go old school, and insert

   #ifdef ARCH_32_BITS // symbol TBD
    int32_t unused;
   #endif

at the end of the struct. Kinda yuck, but would work. I guess.

**Comment by steven-johnson** [comment]
Wait: duh, anonymous unions are a thing, aren't they?

    union {
        /** A pointer to the start of the data in main memory. */
        uint8_t* host;

        /** unused padding, present to ensure that this struct
         * is a multiple of 64 bits on 32-bit systems. */
        uint64_t unused;
    };

**Comment by steven-johnson** [comment]
OK, here's a better question: why is memoize so sensitive to sizeof(buffer_t)? Wouldn't it be simpler to just address that?

**Comment by abadams** [comment]
By "sensitive to" I mean it's a canary. It reliably breaks when the runtime
has a different idea of what buffer_t is to code that uses it.

I was indeed thinking of the anonymous union solution. It didn't work out
of the box - it looks like it would also need a change inside CodeGen_LLVM
where buffer_ts are created.

**Comment by steven-johnson** [comment]
Just pushed a hopeful fix; haven't tested on 32-bit yet but should do the trick

**Comment by zvookin** [comment]
The API here has to go. If we're going to do this, the flag words should be opaque and there should be API calls for the client to say "the host buffer was modified outside of Halide" and "The device buffer was modified outside of Halide". If we want, we could allow querying the flags, but that is only useful for debugging really. halide_set_host_data_modified and halide_set_device_data_modified are what I was thinking of for the buffer_t redesign. (Though it is possible the idea of host and device will go away and instead a buffer will have one memory like thing it points to plus a backing buffer chain.)

**Comment by abadams** [comment]
Yet another solution with a trick suggested by Dillon:
    ...
    bool host_dirty __attribute__((aligned(1)));
    bool dev_dirty __attribute__((aligned(1)));
    uint8_t _padding[10 - sizeof(void *)];
} buffer_t

**Comment by dsharletg** [comment]
I think the right thing to do here is use #pragma pack, and then use another attribute to force the alignment of the struct itself on the compiler that misinterprets pragma pack.

No amount of manually inserted padding members can guarantee that the data layout is the same across all compilers. The only way I can think of to do that is to pragma pack.

Have we tried #pragma pack(4) instead of 1? That might get us a consistent data layout, and avoid the struct itself being unaligned.

**Comment by steven-johnson** [comment]
> The API here has to go... flag words should be opaque

yeah... I think you're right. willfix.

> I think the right thing to do here is use #pragma pack

I stand by my assertion that #pragma pack is the devil and we're better off relying on carefully crafted structures.

> anonymous unions are a thing

after reflection overnight, I'm now thinking that the anonymous-union approach is preferable to the add-an-unused-pointer approach; for one thing, it guarantees a fixed buffer_t size everywhere. any other opinions?

**Comment by abadams** [comment]
One worry I had with the anonymous union - can you still make a buffer_t
like so:

buffer_t buf = {dev, host_ptr, ....};

?

**Comment by steven-johnson** [comment]
On Wed, Apr 8, 2015 at 10:56 AM, Andrew Adams <notifications@github.com>
wrote:

> One worry I had with the anonymous union - can you still make a buffer_t
> like so:
>
> buffer_t buf = {dev, host_ptr, ....};
>

...that is an excellent question; I don't know what C++11 specifies for
that case. Will examine.

**Comment by abadams** [comment]
R.e. flags being opaque: It should either be an abstract data type only touched via an API, or not. Currently it's not. I think the format of all the fields should be part of the interface so that's it's possible to initialize buffer_ts as constants.

I agree the API calls should be more meaningful though, and probably should just mirror Image<T> and Buffer:

halide_buffer_set_host_dirty(buffer_t *buf, bool value);
halide_buffer_set_dev_dirty(buffer_t *buf, bool value);

**Comment by steven-johnson** [comment]
Re: initialization, if you just do 

    buffer_t buf = {dev, host_ptr, ....};

you'll usually get 

    warning: suggest braces around initialization of subobject

which you can avoid by doing

     buffer_t buf = {dev, {host_ptr}, ....};

which may or may not be a deal-breaker. Thoughts?

**Comment by abadams** [comment]
LGTM.

File: src/runtime/HalideRuntime.h [hunk 2]
```diff
@@ -475,21 +474,60 @@ typedef struct buffer_t {
...
-    bool dev_dirty;
-
-    uint8_t _padding[2];
-} buffer_t;
-#pragma pack(pop)
+    halide_buffer_dev_dirty = (1 << 1)
+};
 
-#endif
+inline bool halide_buffer_get_host_dirty(const buffer_t* buf) {
```
Comment (dsharletg) [inline src/runtime/HalideRuntime.h]: Should these be `halide_buffer_is_host_dirty`? Note 'is' vs. 'get'.

**Comment by dsharletg** [comment]
LGTM aside from the super minor naming suggestion. 

In the future, we should consider making all buffer_t members set via a function like this, not just the flags. I think that's the only foolproof/non-compiler-hack way to avoid the issue of data layout.

**Comment by steven-johnson** [comment]
Two LGTM's; any objections? Otherwise I will merge today.

File: src/runtime/HalideRuntime.h [hunk 3]
```diff
@@ -475,21 +474,60 @@ typedef struct buffer_t {
...
+    return (buf->flags & halide_buffer_host_dirty) != 0;
+}
+
+inline bool halide_buffer_get_dev_dirty(const buffer_t* buf) {
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: Given the issue we had this morning with inline on os x, make sure you test apps/glsl

Comment (zvookin) [inline src/runtime/HalideRuntime.h]: Have we considered just making the flags 32-bit ints?

-Z-


On Apr 9, 2015, at 3:58 PM, Andrew Adams <notifications@github.com> wrote:

In src/runtime/HalideRuntime.h
<https://github.com/halide/Halide/pull/737#discussion_r28110775>:

>
> -#endif
> +inline bool halide_buffer_get_host_dirty(const buffer_t* buf) {
> +    return (buf->flags & halide_buffer_host_dirty) != 0;
> +}
> +
> +inline bool halide_buffer_get_dev_dirty(const buffer_t* buf) {

Given the issue we had this morning with inline on os x, make sure you test
apps/glsl

—
Reply to this email directly or view it on GitHub
<https://github.com/halide/Halide/pull/737/files#r28110775>.

Comment (abadams) [inline src/runtime/HalideRuntime.h]: Good thought. We'd still need 4 more bytes of padding in 64-bit mode to
ensure the struct is a multiple of 8 bytes.

**Comment by abadams** [comment]
This PR is obsolete now, right?

**Comment by steven-johnson** [comment]
yep

**Pull request closed without merging**